### PR TITLE
Mock instances

### PR DESCRIPTION
**Pull request by BenJoParadise**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/15201

## Related PRs
List related PRs against other branches:

branch | PR
-----|------
proxy-unsecure-fixes | https://github.com/demisto/content/pull/2958

## Required version of Demisto
x.x.x

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Dependencies
Mention the dependencies of the entity you changed as given from the precommit hooks in checkboxes, and tick after tested them.
- [ ] Dependency 1
- [ ] Dependency 2
- [ ] Dependency 3

## Additional changes
Describe additional changes done, for example adding a function to common server.

### CODE REVIEW

File: Tests/mock_server.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        else:
+            local_log_filepath = os.path.join(
+                '/tmp', os.path.basename(id_to_log_file(self.last_playbook_id, self.record)))
+            remote_log_filepath = os.path.join(self.active_folder, id_to_log_file(self.last_playbook_id, self.record))
+
+            with open(local_log_filepath, 'w+') as log:
+                log.write('STDOUT:\n')
+                log.write(self.process.stdout.read())
+                log.write('\nSTDERR:\n')
+                log.write(self.process.stdout.read())
```
Comment (ronykoz) [inline Tests/mock_server.py]: stderr

File: Tests/mock_server.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+            with open(os.devnull, 'w') as FNULL:
```
Comment (ronykoz) [inline Tests/mock_server.py]: comment

File: Tests/scripts/run_installer_on_instance.sh [hunk 3]
```diff
@@ -22,6 +22,13 @@ sleep 90s
 echo "add instance to known hosts"
 ssh-keyscan -H ${PUBLIC_IP} >> ~/.ssh/known_hosts
 
+# install mitmproxy
```
Comment (ronykoz) [inline Tests/scripts/run_installer_on_instance.sh]: check with DevOps the option to add that to the AMI creation process

File: Tests/scripts/clone_mocks.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+set -e
+
+cat > ~/.ssh/config << EOF
```
Comment (ronykoz) [inline Tests/scripts/clone_mocks.sh]: check with DevOps the option to add that to the AMI creation process

File: Tests/scripts/upload_mocks.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+set -e
+cat >~/.gitconfig << EOF
```
Comment (ronykoz) [inline Tests/scripts/upload_mocks.sh]: check with DevOps the option to add that to the AMI creation process

File: Tests/test_content.py [hunk 6]
```diff
@@ -4,10 +4,14 @@
 import string
 import random
 import argparse
+from subprocess import check_output
+
```
Comment (ronykoz) [inline Tests/test_content.py]: remove redundant space

File: Tests/test_content.py [hunk 7]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
     return test_message
 
 
-def run_test(c, failed_playbooks, integrations, playbook_id, succeed_playbooks,
-             test_message, test_options, slack, CircleCI, buildNumber, server_url, build_name):
-    print '------ Test %s start ------' % (test_message,)
-    # run test
+# TODO: Remove before merge
```
Comment (ronykoz) [inline Tests/test_content.py]: remove

File: Tests/test_content.py [hunk 8]
```diff
@@ -293,6 +388,12 @@ def main():
         print('no integrations are configured for test')
         return
 
+    with open('public_ip', 'rb') as f:
+        public_ip = f.read()
```
Comment (ronykoz) [inline Tests/test_content.py]: add blank line

File: Tests/test_content.py [hunk 9]
```diff
@@ -293,6 +388,12 @@ def main():
...
+    ami = AMIConnection(public_ip)
+    ami.clone_mock_data()  # FUTURE: pull instead of clone
```
Comment (ronykoz) [inline Tests/test_content.py]: if DevOps is changing their part then it should be pull

File: Tests/test_content.py [hunk 10]
```diff
@@ -249,6 +334,15 @@ def load_conf_files(conf_path, secret_conf_path):
     return conf, secret_conf
 
 
+def get_content_branch():
```
Comment (ronykoz) [inline Tests/test_content.py]: use $CIRCLE_BRANCH

File: Tests/test_content.py [hunk 11]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+    proxy.set_folder_primary()
+    return succeed
+
+
+def run_test(c, proxy, ami, failed_playbooks, integrations, unmockable_integrations, playbook_id, succeed_playbooks,
+             test_message, test_options, slack, CircleCI, buildNumber, server_url, build_name):
+    print '------ Test %s start ------' % (test_message,)
+
+    if not integrations:
+        print "No integrations, bypassing mock."
```
Comment (ronykoz) [inline Tests/test_content.py]: remove this print

File: Tests/test_content.py [hunk 12]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+
+# run the test without mocking mechanism.
+def mockless_run(c, failed_playbooks, integrations, playbook_id, succeed_playbooks,
+                 test_message, test_options, slack, CircleCI, buildNumber, server_url, build_name):
     succeed, inc_id = test_integration(c, integrations, playbook_id, test_options)
-    # use results
     if succeed:
         print 'PASS: %s succeed' % (test_message,)
         succeed_playbooks.append(playbook_id)
+        # succeed = demo_playback_test(c, proxy, public_ip, integrations, playbook_id, test_options)
```
Comment (ronykoz) [inline Tests/test_content.py]: remove

File: Tests/test_content.py [hunk 13]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
     else:
         print 'Failed: %s failed' % (test_message,)
         failed_playbooks.append(playbook_id)
-        notify_failed_test(slack, CircleCI, playbook_id, buildNumber, inc_id, server_url, build_name)
+        # notify_failed_test(slack, CircleCI, playbook_id, buildNumber, inc_id, server_url, build_name)
```
Comment (ronykoz) [inline Tests/test_content.py]: enable

File: Tests/test_content.py [hunk 14]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+
+    if not integrations:
+        print "No integrations, bypassing mock."
+        mockless_run(c, failed_playbooks, integrations, playbook_id, succeed_playbooks,
+                     test_message, test_options, slack, CircleCI, buildNumber, server_url, build_name)
+        print '------ Test %s end ------' % (test_message,)
 
+        return
+
+    if has_unmockable_integration(integrations, unmockable_integrations):
```
Comment (ronykoz) [inline Tests/test_content.py]: please note it will be a dict so send the keys only

Comment (ronykoz) [reply Tests/test_content.py]: and make that a one if with or cond

File: Tests/test_content.py [hunk 15]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+        print "Unmockable integrations, bypassing mock."
```
Comment (ronykoz) [inline Tests/test_content.py]: remove print

File: Tests/test_content.py [hunk 16]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+        mockless_run(c, failed_playbooks, integrations, playbook_id, succeed_playbooks,
+                     test_message, test_options, slack, CircleCI, buildNumber, server_url, build_name)
+        print '------ Test %s end ------' % (test_message,)
+
+        return
+
+    set_mock_params(integrations)
```
Comment (ronykoz) [inline Tests/test_content.py]: mock_run function

File: Tests/test_content.py [hunk 17]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+#     print "DEMO: Test {} with playback".format("succeeded" if succeed else "failed")
+#     return succeed, proxy_proc
+
+
+def has_unmockable_integration(integrations, unmockable_integrations):
+    return list(set(x['name'] for x in integrations).intersection(unmockable_integrations))
+
+
+# Configure integrations to work with mock
+def set_mock_params(integrations):
```
Comment (ronykoz) [inline Tests/test_content.py]: docstring

Comment (ronykoz) [reply Tests/test_content.py]: change the name to `configure_proxy_unsecure`

File: Tests/test_content.py [hunk 18]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+    proxy.stop()
+    if succeed:
+        proxy.move_to_primary(playbook_id)
+    proxy.set_folder_primary()
+    return succeed
+
+
+def run_test(c, proxy, ami, failed_playbooks, integrations, unmockable_integrations, playbook_id, succeed_playbooks,
+             test_message, test_options, slack, CircleCI, buildNumber, server_url, build_name):
+    print '------ Test %s start ------' % (test_message,)
```
Comment (ronykoz) [inline Tests/test_content.py]: update to what we discussed about

File: Tests/test_content.py [hunk 19]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+        # use results
+        proxy.stop()
+        if succeed:
+            print 'PASS: %s succeed' % (test_message,)
+            succeed_playbooks.append(playbook_id)
+            print '------ Test %s end ------' % (test_message,)
+
+            return
+
+        print "Test failed with mock, rerunning without mock."
```
Comment (ronykoz) [inline Tests/test_content.py]: print "Test failed with mock, recording new mock."

File: Tests/test_content.py [hunk 20]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+        print "Unmockable integrations, bypassing mock."
+        mockless_run(c, failed_playbooks, integrations, playbook_id, succeed_playbooks,
+                     test_message, test_options, slack, CircleCI, buildNumber, server_url, build_name)
+        print '------ Test %s end ------' % (test_message,)
+
+        return
+
+    set_mock_params(integrations)
+    if not proxy.has_mock_file(playbook_id):
+        print "Mock file does not exist, running without mock."
```
Comment (ronykoz) [inline Tests/test_content.py]: remove

File: Tests/test_content.py [hunk 21]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+    else:
+        print "Running with playback"
```
Comment (ronykoz) [inline Tests/test_content.py]: no need

File: Tests/test_content.py [hunk 22]
```diff
@@ -81,20 +85,101 @@ def update_test_msg(integrations, test_message):
...
+# Configure integrations to work with mock
+def set_mock_params(integrations):
+    for elem in integrations:
+        for param in ('proxy', 'useProxy', 'insecure', 'unsecure'):
+            if param in elem['params']:
+                elem['params'][param] = True
+
+
+# run the test without mocking mechanism.
+def mockless_run(c, failed_playbooks, integrations, playbook_id, succeed_playbooks,
```
Comment (ronykoz) [inline Tests/test_content.py]: update the name

**Comment by BenJoParadise** [comment]
Branch and PR were remade: https://github.com/demisto/content/pull/3044

**Pull request closed without merging**